src/core/unit.h

```diff
@@ -793,6 +793,8 @@ int unit_prepare_exec(Unit *u);
 
 void unit_warn_leftover_processes(Unit *u);
 
+extern bool service_exited_release_console(Unit *u);
```

Comment:
The prototype should be in service.h if the code is in service.c. Then unit.c can #include "service.h".

Reply:
Oh, I had considered about it. But it will expose the detail of Service to Unit, is it ok from the closure of the design? If it is ok, I will push another version.

Thank you very much for the review.

Reply:
Hmm, let's maybe do this properly: let's add a new function pointer to UnitVTable called "needs_console()" or so, which is then implemented by the Service types, but not for the others. Then, let's add a generic function "unit_needs_console()" that calls into the unit-type-specific implementation if the function ptr is set (which it would be for services), and use a generic fallback that just uses exec_context_may_touch_console() as the current code does (which would be used for all other unit types)

With that in place there would then be a single call that tells us whether the unit currently needs a console, and it would be simple, and generic, without losing the type-agnostic handling in unit.c.

As a second thing I'd then add a bool "console_reffed" or so to the Unit object, and set it whenever a unit has reffed n_on_console. That would make the ref counting much more robust, and would simplify the code too.

Does that make sense? @pfliu any chance you can rework this accordingly?